# HTML tags position generation

Con la posición base del primer rectángulo, y aquellas posiciones de los saltos necesarios, calcularemos todos los restantes. Las posiciones en los tags de html generados por GIMP vienen dadas por: 'horizontalizq, verticalarriba, horizontalder, verticalabajo'.

Jerarquía de parámetros: switches contienen grupos de bocas infiniband, cada una con 'col' columnas y 'row' filas. 

Input parameters: 
* sw_col: número de 'columnas de switches'
* sw_row: número de 'filas de switches' 
* gr_col: número de 'columnas de grupos'
* gr_row: número de 'filas de grupos'
* col: número de columnas/grupo
* row: número de filas/grupo 

In [8]:
import numpy as np

## Creación de la clase

In [9]:
class mapeador():
    
    def __init__(self, path_input, path_output, dim, sw_col = 1, sw_row = 1, base = 0, show_pos = True):
        self.path_input = path_input
        self.path_output = path_output
        self.sw_col = sw_col
        self.sw_row = sw_row
        self.gr_col = dim[0]
        self.gr_row = dim[1]
        self.col = dim[2]
        self.row = dim[3]
        self.base = base
        self.show_pos = show_pos
        
        self.importar_map()
        
    def importar_map(self):
        with open(self.path_input, 'r') as file:
            self.imgmap = file.read()
    
    def coordenadas(self):
        #Conseguimos el diccionario con las cordenadas del rectángulo inicial y los saltos necesarios
        coords_dict = {}
        split = [split.split('\" alt=\"') for split in self.imgmap.split('coords=\"')[1:]]
        
        for coords in split:
            alt = coords[1].split('\"  nohref')[0] #Careful: dos espacios entre \" y nohref
            coords_dict[int(alt)] = coords[0]
        
        return coords_dict
        
    def distancias(self):
        #Calculamos las distancias al rectángulo inicial (base)
        coords_dict = self.coordenadas()
        
        dist_dict = {}
        pos_base = [int(pos) for pos in coords_dict[1].split(",")]
        
        for entry in coords_dict.keys():
            if entry == 1:
                continue
            pos = [int(pos) for pos in coords_dict[entry].split(",")]
            dist_dict[entry] = np.subtract(pos,pos_base)
        
        #dist_good = {entry: list(dist_dict[entry][:2]) + list(dist_dict[entry][:2]) for entry in dist_dict.keys()} # Nos quedamos con las dos primeras entradas, y las repetimos.
        
        return dist_dict, pos_base
        
    def todas_posiciones(self):
        #obtenemos un diccionario alt <---> posiciones de los rectángulos
        dist_good, pos_base = self.distancias()
    
        alt = 1
        all_pos = {}
        all_pos[1 + self.base] = np.array(pos_base)

        def calcular_posicion(alt,i,j,k,l,m,n):
            #Calculo del siguiente rectángulo.
            change = np.multiply(i,dist_good[2]) + np.multiply(j,dist_good[3]) #al menos habrá filas y columnas.
            if k:
                change += np.multiply(k,dist_good[self.col * self.row + 1])
            if l: 
                change += np.multiply(l,dist_good[self.col * self.row * self.gr_col + 1])
            if m:
                change += np.multiply(m,dist_good[self.col * self.row * self.gr_col * self.gr_row + 1])
            if n: 
                change += np.multiply(n,dist_good[self.col * self.row * self.gr_col * self.gr_row * self.sw_col + 1])


            pos = np.add(pos_base, change)
            alt += 1
            return alt, pos 

        for n in range(self.sw_row):
            for m in range(self.sw_col):
                for l in range(self.gr_row):
                    for k in range(self.gr_col):
                        for j in range(self.col):
                            for i in range(self.row):
                                if i == 0 and j == 0 and k == 0 and l == 0 and m == 0 and n == 0:
                                    continue
                                alt, pos = calcular_posicion(alt,i,j,k,l,m,n)
                                all_pos[alt + self.base] = pos
        if self.show_pos:
            print(all_pos)
        
        return all_pos
    
    def preparar_output(self):
        #Eliminar los tags que ahora sustituiremos
        if self.base != 0:
            with open(self.path_output, 'r') as file:
                self.imgmap = file.read()
        
        splitmap = self.imgmap.split('\n')
        finalmap = [] 
        if self.base == 0: #borrar lineas en el fichero generado a mano
            for split in splitmap:
                if split.find('<area'):
                    finalmap.append(split)
        else: #incorporar a un fichero más filas a partir de un numero base
            finalmap = splitmap
            
        return finalmap
    
    def generar_y_guardar(self):
        # Insertar los tags en la lista que irá al fichero
        all_pos = self.todas_posiciones()
        finalmap = self.preparar_output()
        
        for entry in all_pos.keys():
            intlist = all_pos[entry].tolist()
            strlist = [str(num) for num in intlist]
            val = ",".join(strlist)
            tag = f'<area shape="rect" coords="{val}" alt="{entry}"  nohref="nohref" />'
            finalmap.insert(-2,tag)
        
        with open(self.path_output, 'w') as file:
            file.write('\n'.join(finalmap))
        
        print("\nalright!")

### Caso 1: switch leafs del switch SX6536

In [10]:
#Detalles switch:
gr_col_1 = 3
gr_row_1 = 1
col_1 = 3
row_1 = 2

In [11]:
#Input/output:
path_input_1 = 'mapeos/[SX6536_detalle] (imported).map'
path_output_1 = 'mapeos/SX6536_detalle.map'

In [12]:
leafSX6536 = mapeador(path_input_1, path_output_1, [gr_col_1, gr_row_1, col_1, row_1])

In [13]:
leafSX6536.generar_y_guardar()

{1: array([108,  61, 141,  76]), 2: array([108,  84, 142,  99]), 3: array([143,  61, 176,  76]), 4: array([143,  84, 177,  99]), 5: array([178,  61, 211,  76]), 6: array([178,  84, 212,  99]), 7: array([219,  61, 251,  76]), 8: array([219,  84, 252,  99]), 9: array([254,  61, 286,  76]), 10: array([254,  84, 287,  99]), 11: array([289,  61, 321,  76]), 12: array([289,  84, 322,  99]), 13: array([330,  61, 361,  76]), 14: array([330,  84, 362,  99]), 15: array([365,  61, 396,  76]), 16: array([365,  84, 397,  99]), 17: array([400,  61, 431,  76]), 18: array([400,  84, 432,  99])}

alright!


### Caso 2: switch SX6036

In [14]:
#Detalles switch:
gr_col_2 = 3
gr_row_2 = 1
col_2 = 6
row_2 = 2

In [15]:
#Input/output:
path_input_2 = 'mapeos/[StorageReview-Mellanox-SX6036-InfiniBand-Switch-Front] (imported).map'
path_output_2 = 'mapeos/SX6036_IB_Front.map'

In [16]:
switchSX6036 = mapeador(path_input_2, path_output_2, [gr_col_2, gr_row_2, col_2, row_2])
switchSX6036.generar_y_guardar()

{1: array([ 70,  51, 105,  67]), 2: array([ 71,  76, 106,  92]), 3: array([108,  51, 142,  67]), 4: array([109,  76, 143,  92]), 5: array([146,  51, 179,  67]), 6: array([147,  76, 180,  92]), 7: array([184,  51, 216,  67]), 8: array([185,  76, 217,  92]), 9: array([222,  51, 253,  67]), 10: array([223,  76, 254,  92]), 11: array([260,  51, 290,  67]), 12: array([261,  76, 291,  92]), 13: array([300,  53, 334,  67]), 14: array([301,  78, 335,  92]), 15: array([338,  53, 371,  67]), 16: array([339,  78, 372,  92]), 17: array([376,  53, 408,  67]), 18: array([377,  78, 409,  92]), 19: array([414,  53, 445,  67]), 20: array([415,  78, 446,  92]), 21: array([452,  53, 482,  67]), 22: array([453,  78, 483,  92]), 23: array([490,  53, 519,  67]), 24: array([491,  78, 520,  92]), 25: array([530,  55, 563,  67]), 26: array([531,  80, 564,  92]), 27: array([568,  55, 600,  67]), 28: array([569,  80, 601,  92]), 29: array([606,  55, 637,  67]), 30: array([607,  80, 638,  92]), 31: array([644,  5

### Caso 3: switch SX6536 (parte 1)

Vamos a dividir el switch SX6536 en dos partes, ya que hay una separación con una barra de metal entre una parte con 12 switches encima de una parte con 24 (en ambos casos, en dos columnas). Esta irregularidad en la cantidad de switches existente en ambas partes hace dificil automatizarlo. Para no comerme la cabeza, simplemente he añadido un parámetro 'base' a partir del cual empezar a contar, en este caso el segundo grupo de switches comienza en el 217 (ese sería el 1, por lo que la base es 216).

In [22]:
#Detalles switch:
sw_col_3 = 2
gr_col_3 = 3
gr_row_3 = 6
col_3 = 3
row_3 = 2

In [23]:
#Input/output:
path_input_3 = 'mapeos/[sx6536] (imported).map'
path_output_3 = 'mapeos/sx6536_full.map'

In [24]:
switchSX6036 = mapeador(path_input_3, path_output_3, [gr_col_3, gr_row_3, col_3, row_3], sw_col = sw_col_3, show_pos = False)
switchSX6036.generar_y_guardar()


alright!


### Caso 3: switch SX6536 (parte 2)

In [25]:
#Detalles switch:
sw_col_32 = 2
gr_col_32 = 3
gr_row_32 = 12
col_32 = 3
row_32 = 2
base = 216

In [26]:
#Input/output:
path_input_32 = 'mapeos/[sx6536] (imported) - 2.map'
path_output_32 = 'mapeos/sx6536_full.map'

In [28]:
switchSX6036_2 = mapeador(path_input_32, path_output_32, [gr_col_32, gr_row_32, col_32, row_32], sw_col = sw_col_32, base = base, show_pos = False)
switchSX6036_2.generar_y_guardar()


alright!


### Caso 4: CPD 1
Esto no es un switch pero bueno, es lo que hay.

In [44]:
#Detalles
gr_col_4 = 1
gr_row_4 = 1
col_4 = 10
row_4 = 2

In [45]:
#Input/output:
path_input_4 = 'mapeos/cpd1_inicial.map'
path_output_4 = 'mapeos/cpd1_mapeo.map'

In [62]:
cpd1 = mapeador(path_input_4, path_output_4, [gr_col_4, gr_row_4, col_4, row_4], show_pos = True)
cpd1.generar_y_guardar()

{1: array([211, 165, 323, 278]), 2: array([212, 529, 323, 641]), 3: array([333, 166, 443, 278]), 4: array([334, 530, 443, 641]), 5: array([455, 167, 563, 278]), 6: array([456, 531, 563, 641]), 7: array([577, 168, 683, 278]), 8: array([578, 532, 683, 641]), 9: array([699, 169, 803, 278]), 10: array([700, 533, 803, 641]), 11: array([821, 170, 923, 278]), 12: array([822, 534, 923, 641]), 13: array([ 943,  171, 1043,  278]), 14: array([ 944,  535, 1043,  641]), 15: array([1065,  172, 1163,  278]), 16: array([1066,  536, 1163,  641]), 17: array([1187,  173, 1283,  278]), 18: array([1188,  537, 1283,  641]), 19: array([1309,  174, 1403,  278]), 20: array([1310,  538, 1403,  641])}

alright!


### Caso 5: CPD 2
Lo mismo de antes

In [66]:
#Detalles
gr_col_5 = 1
gr_row_5 = 1
col_5 = 12
row_5 = 2

In [67]:
#Input/output:
path_input_5 = 'mapeos/cpd2_inicial.map'
path_output_5 = 'mapeos/cpd2_mapeo.map'

In [68]:
cpd2 = mapeador(path_input_5, path_output_5, [gr_col_5, gr_row_5, col_5, row_5], show_pos = True)
cpd2.generar_y_guardar()

{1: array([ 90, 313, 214, 440]), 2: array([ 90, 581, 214, 706]), 3: array([225, 313, 349, 440]), 4: array([225, 581, 349, 706]), 5: array([360, 313, 484, 440]), 6: array([360, 581, 484, 706]), 7: array([495, 313, 619, 440]), 8: array([495, 581, 619, 706]), 9: array([630, 313, 754, 440]), 10: array([630, 581, 754, 706]), 11: array([765, 313, 889, 440]), 12: array([765, 581, 889, 706]), 13: array([ 900,  313, 1024,  440]), 14: array([ 900,  581, 1024,  706]), 15: array([1035,  313, 1159,  440]), 16: array([1035,  581, 1159,  706]), 17: array([1170,  313, 1294,  440]), 18: array([1170,  581, 1294,  706]), 19: array([1305,  313, 1429,  440]), 20: array([1305,  581, 1429,  706]), 21: array([1440,  313, 1564,  440]), 22: array([1440,  581, 1564,  706]), 23: array([1575,  313, 1699,  440]), 24: array([1575,  581, 1699,  706])}

alright!
